server/src/main/java/org/apache/druid/segment/realtime/firehose/EventReceiverFirehoseFactory.java

```diff
@@ -350,34 +427,45 @@ public long getBytesReceived()
       return bytesReceived.get();
     }
 
+    /**
+     * This method is synchronized because it might be called concurrently from multiple threads: from {@link
+     * #delayedCloseExecutor}, and explicitly on this Firehose object.
+     */
     @Override
-    public void close()
+    public synchronized void close()
     {
-      if (!closed) {
-        log.info("Firehose closing.");
-        closed = true;
+      if (closed) {
+        return;
+      }
+      closed = true;
```

Comment:
Not sure what do you mean. `synchronized (this)` doesn't throw any exceptions

Reply:
For example `eventReceiverFirehoseRegister.unregister(serviceName);` throws an exception, then the `close()` is not complete, but the next calls to `close()` won't try again as `closed == true` already.

Reply:
If it throws an exception once, it will likely throw it the second time as well. I think the general practice is to set such flags up front and don't allow to "close again" in case of errors. In user code, when close() throws an exception, nobody tries to close an object again either.

Reply:
I think general practice is to set `closed` flag when the closing is successful. At least for reference `ServerSocket.close()`:
```java
    public void close() throws IOException {
        synchronized(closeLock) {
            if (isClosed())
                return;
            if (created)
                impl.close();
            closed = true;
        }
    }
```

Reply:
I checked Druid sources (search `closed = true`) and it's a random mix of closing first and closing last. In `EventReceiverFirehose` closing last gives nothing but the possibility of a Thread leak (delayedCloseExecutor may not exit).

Reply:
However, I was wrong, there is no possibility for a Thread leak, `delayedCloseExecutor` should exit anyway. But I don't see a point in moving `closed = true;` statement to the end of the method either. If you disagree perhaps you could open an issue for aligning to the same practice (`closed = true` in the end) across all Druid classes. If more people support support this stance then appropriate changes could be made.

Reply:
> I checked Druid sources (search `closed = true`) and it's a random mix of closing first and closing last.

Got it.